pkg/scheduler/queue/query_component_capacity.go

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"math"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+)
+
+type QueryComponent string
+
+const (
+	StoreGateway QueryComponent = "store-gateway"
+	Ingester     QueryComponent = "ingester"
+)
+
+// cannot import constants from frontend/v2 due to import cycle
+// these are attached to the request's AdditionalQueueDimensions by the frontend.
+const ingesterQueueDimension = "ingester"
+const storeGatewayQueueDimension = "store-gateway"
+const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
+
+// QueryComponentNameForRequest parses the expected query component from annotations by the frontend.
+func QueryComponentNameForRequest(req *SchedulerRequest) string {
```

Comment:
I find the name "component name" very confusing given it could contain multiple components. Perhaps "affected components" would be clearer? 

(same feedback applies for `ExceedsCapacityForComponentName` below - I think something like `ExceedsCapacityForAffectedComponents` might be clearer)

Reply:
The fundamental problem is that we have a single string that represents both being passed from the front end, and we have to call that something.

We also need a name for the separate components themselves, as well as these flags that translate from the single-string component name to when a given component used, like `isStoreGateway` and `isIngester`.

I don't use "Components" because that seems to indicate multiple of our type, `QueryComponent`.

I have tried to disambiguate a bit here by using:
* single string: `ExpectedQueryComponentName` / `ComponentName`
* flags: `Flags`
* individual components: the `QueryComponent` type

I think ultimately the solutino here is the solution for everything else, which is fix it in the frontend.

We can abandon this:
`const ingesterAndStoreGatewayQueueDimension = "ingester-and-store-gateway"`

for just something like

```golang
type ExpectedQueryComponents struct {
    ingester bool
    storeGateway bool
}
```
then this whole concept of "something that is a single value but indicates multiple things" is gone

Reply:
>I think ultimately the solution here is the solution for everything else, which is fix it in the frontend.

That sounds good to me, but I'm happy for that to come in a separate PR.